### PR TITLE
Add SAM DMA request data to device files

### PR DESCRIPTION
**Pull request by chris-durand**

### CODE REVIEW

**Comment by chris-durand** [comment]
@salkinium The CI and my local python disagree on the order of 2 lines in one device file.
```
+      <source device-flash="17" device-variant="d|l" name="Fdpll" value="8"/>
       <source device-flash="15|16" device-variant="a|b|c|l" name="Dpll96m" value="8"/>
       <source device-flash="17|18" device-variant="a" name="Dpll96m" value="8"/>
-      <source device-flash="17" device-variant="d|l" name="Fdpll" value="8"/>
```
Is there anything I should do about it?

**Comment by salkinium** [comment]
Is the data up-to-date? I know there are some corner cases where the tree overlay algorithm isn't working correctly, but I broke my brain trying think about it, so it's not fixed…

**Comment by salkinium** [comment]
Hm, also the CI is running `ubuntu-latest` and Python version `3.x`, so that is kind of a moving target and we should nail that down a bit more to avoid any kind of unclear regressions like this.

**Comment by chris-durand** [comment]
> Is the data up-to-date? I know there are some corner cases where the tree overlay algorithm isn't working correctly, but I broke my brain trying think about it, so it's not fixed…

Yes, must be up-to-data since it's a fresh repo. It has just downloaded it.

**Comment by salkinium** [comment]
I think they disagree because your code is missing the sort key. Otherwise the tree merger doesn't know what attributes to use for sorting.

File: tools/generator/dfg/sam/sam_device_tree.py [hunk 1]
```diff
@@ -287,6 +294,20 @@ def driverOrder(e):
                     source.setAttributes("name", name, "value", source_id)
                 generators = driver.addChild("generators")
                 generators.setAttributes("value", str(p["gclk_data"]["generator_count"]))
+            # Add request data to DMA module
+            # Skip D09 devices, information is missing in raw data.
+            elif name in ("dmac", "xdmac", "pdc") and p["id"].series != "d09":
+                requests = driver.addChild("requests")
+                requests.addSortKey(lambda e: int(e["id"]))
```
Comment (salkinium) [inline tools/generator/dfg/sam/sam_device_tree.py]: Oh wait, it's here… hmmm

File: tools/generator/dfg/sam/sam_device_tree.py [hunk 2]
```diff
@@ -268,7 +276,6 @@ def driverOrder(e):
                     inst = driver.addChild("instance")
                     inst.setValue(i[len(dtype):])
             if name == "gclk":
-                instance_pattern = re.compile(r"^(?P<per>([a-z]+))(?P<instance>([0-9]+))$")
                 driver.addSortKey(lambda e: (e.name, int(e["value"])))
```
Comment (salkinium) [inline tools/generator/dfg/sam/sam_device_tree.py]: This key is incomplete though. There are several child node that have the same name and value in the gclk node.

Comment (salkinium) [reply tools/generator/dfg/sam/sam_device_tree.py]: Let's see if that fixed it.

File: tools/generator/dfg/sam/sam_device_tree.py [hunk 3]
```diff
@@ -276,7 +276,7 @@ def driverOrder(e):
                     inst = driver.addChild("instance")
                     inst.setValue(i[len(dtype):])
             if name == "gclk":
-                driver.addSortKey(lambda e: (e.name, int(e["value"])))
+                driver.addSortKey(lambda e: (e.name, int(e["value"]), e.get("peripheral", ""), e.get("instance", ""), e.get("name", "")))
```
Comment (chris-durand) [inline tools/generator/dfg/sam/sam_device_tree.py:279]: ~I guess that could be the right fix, but this is the wrong place. That is the sort key for the GCLK data, not the DMA requests :wink:.~ Oh sorry, there is a different issue now. Your fix has indeed solved my original problem but the CI shows another. Thanks.

File: tools/generator/dfg/sam/sam_device_tree.py (hunk 1 above)
Comment (chris-durand) [reply tools/generator/dfg/sam/sam_device_tree.py]: There is the same issue now for the DMA data:
```
+        <request device-variant="b" signal="ch0_tx" id="30" peripheral="dacc"/>
         <request device-series="e70|s70" device-pin="n" device-flash="21" device-variant="a" signal="tx" id="30" peripheral="dacc"/>
         <request device-series="e70|s70|v70" device-pin="j|n|q" device-flash="19|20" device-variant="a" signal="tx" id="30" peripheral="dacc"/>
         <request device-series="e70|s70|v71" device-pin="j|q" device-flash="20|21" device-variant="a" signal="tx" id="30" peripheral="dacc"/>
         <request device-series="v71" device-pin="n|q" device-flash="19" device-variant="a" signal="tx" id="30" peripheral="dacc"/>
-        <request device-variant="b" signal="ch0_tx" id="30" peripheral="dacc"/>
```

All these lines represent the entry `signal="ch0_tx" id="30" peripheral="dacc"`, but they differ in the filters added by the generator (`device-series`, `device-pin`, `device-flash`, `device-variant`). These would need to be included in the sort key somehow.

Comment (salkinium) [reply tools/generator/dfg/sam/sam_device_tree.py]: The tree overlay algorithm is order deterministic, therefore there is no extra sort key necessary.
I refactored the requests a little (no need for a plural parent node) and added a more extensive sort key, seems to have fixed it. If you're happy please rebase the fixups.

Comment (chris-durand) [reply tools/generator/dfg/sam/sam_device_tree.py]: Thanks. I could have sworn I tested the same sort key change you added yesterday and it did not work for me. Must have messed something else up. I am happy it works now.

**Review by salkinium — APPROVED** [review]
Thank you!

**Pull request merged**